vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,468 @@
+/************************************************************************
+ * This file introduces a class to model a Routing Resource Graph (RRGraph or RRG) 
+ * which is widely used by placers, routers, analyzers etc.
+ * A Routing Resource Graph (RRGraph or RRG) consists of a number of nodes and edges.
+ *
+ * Each node represents a routing resource, which could be 
+ * 1. a routing track in X-direction or Y-direction (CHANX or CHANY) 
+ * 2. an input or an output of a logic block (IPIN or OPIN) 
+ * 3. a virtual source or sink node (SOURCE or SINK), which are starting/ending points of routing trees.
+ *
+ * Each edge represents a switch between routing resources, which could be
+ * 1. a multiplexer
+ * 2. a tri-state buffer
+ * The switch information are categorized in the rr_switch_inf of RRGraph class.
+ *
+ * IMPORTANT:
+ *   The RRGraph is designed to be a read-only database/graph, once created.
+ *   Placement and routing should not change any attributes of RRGraph.
+ *   Any placement and routing results should be stored in other data structures, such as PlaceContext and RoutingContext. 
+ *
+ * How to use the RRGraph data structure 
+ * =====================================
+ * To iterate over the nodes/edges/switches/segments in a RRGraph, 
+ *    using a range-based loop is suggested.
+ *  -----------------------------------------------------------------
+ *    Example: iterate over all the nodes
+ *      for (RRNodeId node : nodes()) {
+ *        // Do something with each node
+ *      }
+ *
+ *      for (RREdgeId edge : edges()) {
```

Comment:
Rather than iterating through all the main components, I'd instead show which illustrates how to walk and trace cross-references. Perhaps something like walking all nodes and for each printing out what downstream nodes they connect to.